feat: Create Camera component

CHAERYEON-LEE opened the pull request:
### 1️⃣ PR 타입(하나 이상의 PR 타입을 선택해주세요)
- [x] 기능 추가
- [ ] 기능 삭제
- [ ] 버그 수정
- [ ] 의존성, 환경 변수, 빌드 관련 코드 업데이트
<br>

### 2️⃣ 이슈 번호
ex) close #19 
<br>
<br>

### 3️⃣ 변경 사항
카메라 컴포넌트 완성, 사진찍기 구현 완료
- 처음 렌더링 됐을 때 사진 찍으려면 회색 화면 눌러서 사진 찍기
- 두번째 사진찍을때부터는 다시찍기 버튼 누르기
- 사진찍을때는 저장 안되고, 완료 버튼을 눌렀을 때만 base64 URL 생성
- 좌우반전해서 사진 저장
- 3,2,1 카운트다운 후 사진찍히도록 구현함
- device마다 가로세로 화면 비율이 달라서 캔버스에 그렸을 때 세로가 늘어난 비율로 그려짐(미해결)
- 스토리북 테스트 완료
<br>
<br>

CHAERYEON-LEE committed:
feat: Create Camera component

카메라 컴포넌트 완성
- 처음 렌더링 됐을 때 사진 찍으려면 회색 화면 눌러서 사진 찍기
- 두번째 사진찍을때부터는 다시찍기 버튼 누르기
- 사진찍을때는 저장 안되고, 완료 버튼을 눌렀을 때만 base64 URL 생성
- 좌우반전해서 사진 저장
- 3,2,1 카운트다운 후 사진찍히도록 구현함
- device마다 가로세로 화면 비율이 달라서 캔버스에 그렸을 때 세로가 늘어난 비율로 그려짐(미해결)
- 스토리북 테스트 완료

pereng11 commented on src/components/common/Button/Button.tsx:
```diff
@@ -9,10 +9,18 @@ export type ButtonProps = {
+  /**
+   * 버튼의 onClick 이벤트를 줄 수 있습니다.
+   */
+  onClick?: () => void;
```
onClick 좋습니다!
그런데 반환값이 void가 아닌 경우도 있어서 일단은 any로 주면 더 좋을 것 같습니다!

CHAERYEON-LEE replied:
오오오 넵!!!

pereng11 commented on src/components/common/Camera/Camera.tsx:
```diff
@@ -0,0 +1,132 @@
+    console.log(canvasRef.current?.toDataURL('image/webp', 0.8));
+  };
+
+  const countDownAndTakeAPicture = () => {
+    const setCount = setInterval(() => {
+      setCountdown((prev) => {
+        if (prev === 1) {
+          clearInterval(setCount);
+          takePhoto();
+        }
+        return prev - 1;
+      });
+    }, 1000);
+    setTimeout(() => {
+      takePhoto();
+      setIsCounting(false);
+    }, 3000);
+  };
```
카운트 다운이 끝났을 때, 값을 0으로 두지 말고 초기값으로 되돌리는 건 어떤가요?
그러면  handlerReplayPicture를 좀 더 깔끔하게 쓸 수 있을 것 같습니다!

CHAERYEON-LEE replied:
countdown 컨테이너에 
조건문으로 countdown === 0 일때는 숨겨놓고, 나중에 초기값 3으로 세팅했을때 다시 보여지도록 조건을 줘서 
바로 초기값을 넣어주는 부분이 안될것같아요ㅠㅠ 
지금은 countdown === 0말고는 적당한 조건이 생각나지 않는데 생각해볼게욥!!!

pereng11 commented on src/components/common/Camera/Camera.tsx:
```diff
@@ -0,0 +1,132 @@
+  const videoRef = useRef<HTMLVideoElement>(null);
+
+  const [isFirstClick, setIsFirstClick] = useState(false);
+  const [isCounting, setIsCounting] = useState(false);
```
countdown 상태값을 활용하면 이 변수를 useRef로 대체할 수도 있을 것 같습니다
화면에 그릴 countdown값이 1,2,3일 경우에 초기값을 4로 주고 countdown값이 3 이하인 경우를 기점으로 조건을 주는 방법도 가능할 것 같고, 다른 더 좋은 방법도 있을 것 같아요!

CHAERYEON-LEE replied:
isFirstClick 상태는 처음 클릭했냐 아니냐를 판단해야해서 꼭 있어야 할 것 같구 
isCounting은 countdown 상태값으로 판단하는 조건문을 추가해서 
isCounting 상태는 지웠어요ㅎㅎ!!👍

pereng11 commented on src/components/common/Camera/Camera.tsx:
```diff
@@ -0,0 +1,132 @@
+    console.log(canvasRef.current?.toDataURL('image/webp', 0.8));
+  };
+
+  const countDownAndTakeAPicture = () => {
+    const setCount = setInterval(() => {
+      setCountdown((prev) => {
+        if (prev === 1) {
+          clearInterval(setCount);
+          takePhoto();
+        }
+        return prev - 1;
+      });
+    }, 1000);
+    setTimeout(() => {
+      takePhoto();
+      setIsCounting(false);
+    }, 3000);
```
지금 handlerReplayPicture함수에서 setIsCounting이 하는 일을 이 함수로 옮기는 건 어떤가요?
카운팅 시작한다 -> 숫자 내린다 -> 다 내리면 사진 찍고 초기값으로 되돌린다
이런 흐름도 괜찮아보입니다

CHAERYEON-LEE replied:
countdown === 0 기준인 조건문이 있어서 바로 초기값으로 돌리면 안될것같아요ㅠㅠ
카운팅 시작한다 -> 숫자 내린다 -> 1이되면 사진찍고 0인상태로 대기한다 -> 사진을 찍으려면 다시 countdown을 3으로 셋팅한다
이런식으로 지금 하고있어욥

pereng11 commented on src/components/common/Camera/Camera.tsx:
```diff
@@ -0,0 +1,132 @@
+    }, 3000);
+  };
+
+  const handlerPlayPicture = () => {
+    setIsFirstClick(true);
+    setDevice();
+    countDownAndTakeAPicture();
+  };
+
+  const handlerReplayPicture = () => {
+    if (!isFirstClick) return;
+    setIsCounting(true);
+    setCountdown(3);
+    countDownAndTakeAPicture();
+  };
+
+  const handlerSavePhoto = () => {
+    if (!isFirstClick) return;
+    savePhoto();
+  };
```
1. handle+명사 또는 작업+handler 형식으로 쓰는 게 약간 국룰인 것 같더라구요
2. play/replay는 사진을 찍는다는 역할과는 조금 맞지 않는 것 같습니다. 위에서 takePhoto라는 이름을 써버려서 그런 것 같은데 혹시 맞나요? 그래도 좀 더 고민해보면 좋을 것 같습니다!

CHAERYEON-LEE replied:
오 handler 이름 전부 수정하겠습니다!!!ㅎㅎ

pereng11 commented on src/components/common/Camera/Camera.tsx:
```diff
@@ -0,0 +1,132 @@
+  return (
+    <div className={styles.container}>
+      <div className={styles.view}>
+        <div
+          className={`${styles.cameraTextContainer} ${
+            isFirstClick ? styles.hidden : ''
+          }`}
+          onClick={handlerPlayPicture}
+          aria-hidden="true"
+        >
+          <p className={styles.cameraTitle}>히어로 사진 찍기</p>
+          <p className={styles.plusText}>+</p>
+        </div>
+        <div
+          className={`${styles.countdownContainer} ${
+            countdown === 0 ? styles.hidden : styles.show
+          }`}
+        >
```
상태에 따라 달라지는 렌더링 여부를 css visibility로 제어하신 이유가 궁금합니다
저는 여기에서 바로 isFirstClick ? 그린다 : 안그린다로 처리하는 게 더 선언적인 것 같아서요!

CHAERYEON-LEE replied:
이부분 제가 고쳐보려구했는데 css 로 배치를 다 해놔서ㅠㅜㅠㅠ 자스로 아예 없앴다 나타났다 이렇게 조절하다보니 
레이아웃이 자꾸 틀어져서요 흑흑..

pereng11 reviewed (commented):
저는 다뤄보지 않은 기술이라서 궁금하기도 하고 걱정도 됐는데 너무 잘해주신 것 같아요! 
aria 속성도 꼼꼼하게 들어가고 카운트다운같은 디테일까지 미쳤습니다~!~! 👍 👍

CHAERYEON-LEE committed:
fix: PR comment amend complete

PR 코멘트 수정 완료
- isCounting state 제거
- isCounting 상태에 의존하고 있던 조건 countdown 상태로 처리하도록 조건문 수정
- canvas 비율 안변하게 캡쳐하기 완료(기존에 위아래로 늘어나던것 수정
- countDownAndTakeAPicture 함수 불필요한 setTimeout 제거
- handler 이름 변경

pereng11 reviewed (approved):
👍
